Source/NSFontCollection.m

```diff
@@ -0,0 +1,341 @@
+                                   visibility: (NSFontCollectionVisibility)visibility
+{
+  return [__sharedFontCollections objectForKey: name];
+}
+
+// Descriptors
+- (NSArray *) queryDescriptors  // copy
+{
+  return [_queryDescriptors copy];
+}
+
+- (NSArray *) exclusionDescriptors
+{
+  return [_exclusionDescriptors copy];
+}
+
+- (NSArray *) matchingDescriptors
+{
+  return [self matchingDescriptorsWithOptions: nil];
+}
+
+- (NSArray *) matchingDescriptorsWithOptions: (NSDictionary *)options
+{
+  GSFontEnumerator *fen = [GSFontEnumerator sharedEnumerator];
+  return [fen matchingFontDescriptorsFor: options];
+}
+
+- (NSArray *) matchingDescriptorsForFamily: (NSString *)family
+{
+  return [self matchingDescriptorsForFamily: family options: nil];
+}
+
+- (NSArray *) matchingDescriptorsForFamily: (NSString *)family options: (NSDictionary *)options
+{
+  NSMutableArray *r = [NSMutableArray arrayWithCapacity: 50];
+  NSArray *a = [self matchingDescriptorsWithOptions: options];
+  NSEnumerator *en = [a objectEnumerator];
+  id o = nil;
+
+  while((o = [en nextObject]) != nil)
+    {
+      if ([[o familyName] isEqualToString: family])
+        {
+          [r addObject: o];
+        }    
+    }
+  
+  return [a copy];
+}
+
+- (instancetype) copyWithZone: (NSZone *)zone
+{
+  NSFontCollection *fc = [[NSFontCollection allocWithZone: zone] init];
+
+  ASSIGNCOPY(fc->_fonts, _fonts);
+  ASSIGNCOPY(fc->_queryDescriptors, _queryDescriptors);
+  ASSIGNCOPY(fc->_exclusionDescriptors, _exclusionDescriptors);
+  ASSIGNCOPY(fc->_queryAttributes, _queryAttributes);
+
+  return fc;
+}
+
+- (instancetype) mutableCopyWithZone: (NSZone *)zone
+{
+  NSMutableFontCollection *fc = [[NSMutableFontCollection allocWithZone: zone] init];
+
+  [fc _setFonts: _fonts];
+  [fc setQueryDescriptors: _queryDescriptors];
+  [fc setExclusionDescriptors: _exclusionDescriptors];
+  [fc _setQueryAttributes: _queryAttributes];
+
+  return fc;
+}
+
+- (instancetype) initWithCoder: (NSCoder *)coder
+{
+  self = [super init];
+  if (self != nil)
+    {
+      if ([coder allowsKeyedCoding])
+        {
+        }
+    }
+  return self;
+}
+
+- (void) encodeWithCoder: (NSCoder *)coder
+{
+}
+  
+@end
+ 
+
+@implementation NSMutableFontCollection 
+
++ (NSMutableFontCollection *) fontCollectionWithDescriptors: (NSArray *)queryDescriptors
+{
+  NSMutableFontCollection *fc = [[NSMutableFontCollection alloc] init];
+  ASSIGNCOPY(fc->_queryDescriptors, queryDescriptors);
+  return fc;
+}
+
++ (NSMutableFontCollection *) fontCollectionWithAllAvailableDescriptors
+{
+  return [self fontCollectionWithDescriptors:
+                 [[GSFontEnumerator sharedEnumerator] availableFontDescriptors]];
+}
+
++ (NSMutableFontCollection *) fontCollectionWithLocale: (NSLocale *)locale
+{
+  return [self fontCollectionWithAllAvailableDescriptors];
+}
+
++ (NSMutableFontCollection *) fontCollectionWithName: (NSFontCollectionName)name
+{
+  return [__sharedFontCollections objectForKey: name];
```

Comment:
This will not suddenly return a mutable instance. You should call the super implementation and make a mutable copy. The same is true for more of these methods.